File: app/models/card.py
from .db import db

class Card(db.Model):
  __tablename__ = 'cards'

  id = db.Column(db.Integer, primary_key=True, nullable=False)

  card_name = db.Column(db.String(100), nullable=False, unique=True)

  card_image = db.Column(db.String(255), nullable=False)

  mana_cost = db.Column(db.String(100), nullable=False, unique=True)

  cmc = db.Column(db.Integer, nullable=False)

  type_line = db.Column(db.String(100), nullable=False)

  oracle_text = db.Column(db.String(100), nullable=False, unique=True)

  power = db.Column(db.Integer)

  toughness = db.Column(db.Integer)

  colors = db.Column(db.String(100), nullable=False)

  color_identity = db.Column(db.String(100), nullable=False)

  legalities = db.Column(db.String(256), nullable=False)

  def to_dict(self):
            return {
                'id': self.id,
                'card_name': self.card_name,
                'card_image': self.card_image,
                'mana_cost': self.mana_cost,
                'cmc': self.cmc,
                'type_line': self.type_line,
                'oracle_text': self.oracle_text,
                'power': self.power,
                'toughness': self.toughness,
                'colors': self.colors,
                'color_identity': self.color_identity,
                'legalities': self.legalities,
            }

File: react-app/src/store/card.js
const ADD_CARD = 'card/ADD_CARD'
const GET_CARD = 'card/GET_CARD'

const addCard = (card) => {
  return{
    type: ADD_CARD,
    payload: card
  }
}

const getCard = (card) => {
  return{
    type: GET_CARD,
    payload: card
  }
}

export const getOneCard = (id) => async (dispatch) => {
  const res = await fetch(`/api/cards/${id}`)
  const data = await res.json()
  dispatch(getCard(data.card))
}

export const addACard = (cardName) => async (dispatch) => {
  const card = await fetch(`https://api.scryfall.com/cards/named?fuzzy=${cardName}`)
  // .then(response => response.json())
  // .then(data => console.log(data));


  const data = await card.json()

  const card_name = data.name
  const card_image = data.image_uris['small']
  const mana_cost = data.mana_cost
  const cmc = data.cmc
  const type_line = data.type_line
  const oracle_text = data.oracle_text
  const power = data.power
  const toughness = data.toughness
  const colors = `${data.colors}`
  const color_identity = `${data.color_identity}`
  const legalities = data.legalities
  const usableLegalities = JSON.stringify(legalities)


  const res = await fetch('/api/cards/', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      card_name,
      card_image,
      mana_cost,
      cmc,
      type_line,
      oracle_text,
      power,
      toughness,
      colors,
      color_identity,
      legalities: usableLegalities
    })
  })

  if (res.ok) {
    const data = await res.json()
    dispatch(addCard(data))
    window.alert('Card successfully added!')
    return data
  }
  else if (res.status < 500) {
    const data = await res.json()
    if (data.errors) return data.errors
  }
  else {
    return ['An error occurred. Please try again']
  }
}

const cardReducer = (state = {}, action) => {
  let newState;
  switch (action.type) {
    case ADD_CARD:
      newState = {...state, ...action.payload}
      return newState
    case GET_CARD:
      newState = {...state, ...action.payload}
      return newState
    default:
      return state
  }
}

export default cardReducer;

